### ISSUE TITLE
How does one persist the internal memory cache of processed Mock<T> types?

### ISSUE THREAD

**Issue by Jothay** [devlooped/moq#1304]
### Description

Hi, I have a large solution with over 4800 integration tests with heavy use of the Moq framework. After some seriously in-depth checking of the performance, I've isolated one significant road hump. Deep down in Moq, when it performs a get call for the `Object` property of a `Mock<T>` instance, it builds an internal type cache entry for the `T` if it doesn't already have one. So the first run of the get call takes some amount of time, which includes creating lock(s) for thread synchronization and then when it's done, releases the lock(s) and returns the value.
* Calls which occur for the same `typeof(T)` or (different types from as well from what I can tell) at the same time on different threads will cause escalation on the locks, increasing the time required to finish.
* Subsequent calls for mocks of the same `typeof(T)` after the initial locks have cleared up are greatly improved performance because of the internal memory caching, instead of potentially waiting 3 seconds for a response (see dotTrace screenshot), it takes a few nanoseconds/milliseconds to reply.

I have reworked my proof of concept xUnit test case to isolate these type calls to the beginning of the run with a Constructor that just loops all those expected types sequentially. This not only reduced the sync locks deep in Moq but made all the tests that run after it in the same class run without the overhead inside their test (better reporting of time taken per test).

Unfortunately, separate classes run in separate app domains in xUnit, so I can only gain this performance boost once per class. If I were to put all the tests into a single class, I would gain the perf boost from this for all the tests in question, but I would lose the ability to run the tests concurrently with test runners, which is a net loss.

Is there a package or setting or some other special configuration I can perform with Moq 4.x (we're migrating from .NET 4.7.2 to .NET 6+ mid next year) that would tell it to cache that information in a persistent location, like a json file or a separate memory domain that each of the xunit class app domains can access? I'm fine with placing the front-loading trigger into the constructor of every class so that whatever xunit decides to run first is the one that gets the hit but then share the perf gain throughout the rest of the class runs. Note that these are all in a single project, so there's only 1 dll output to run the tests from.

### dotTrace Snapshots
Before isolation, these two examples were running at the same time which escalated the locks (there are ~400 types, about 16 threads would have been going in FIFO to proceess them all). The total time taken would have been in the neighborhood of 3 minutes.
![image](https://user-images.githubusercontent.com/7400141/205172093-9a4e1d16-3c33-473d-91a7-4089cb0b7410.png)
![image](https://user-images.githubusercontent.com/7400141/205175669-838752f4-477b-4433-befa-8e02093679b8.png)
After isolation, I'm front-loading all that cost to a single function. The total time taken to do this in contrast without escalating the locks anymore is about 1 minute 33 seconds.
![image](https://user-images.githubusercontent.com/7400141/205175878-4aea0c39-bde2-4842-af92-436ca0fe47c6.png)
And then the subsequent calls get to be really fast
![image](https://user-images.githubusercontent.com/7400141/205175827-88d6ff94-c984-4618-95f6-f076b1cac79c.png)

### Outpupt from Proof of Concept running in separate classes, with isolation
I added `**` around the relevant time span points so you can find them more easily
```
2022-12-01T15:47:56.6943060-06:00 Start CONSTRUCTOR 1 (with preload)
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:00:00.0234032 Before Run 1 CONSTRUCTOR 1
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:**01:35**.5649612 After Run 1 CONSTRUCTOR 1
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:01:35.5720070 After Run 2 CONSTRUCTOR 1
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:01:35.5740155 After Run 3 CONSTRUCTOR 1

2022-12-01T15:47:56.6943060-06:00 CONSTRUCTOR 2 (No preload)

2022-12-01T15:47:56.6943060-06:00 CONSTRUCTOR 3 (with preload)
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:00:00.0234032 Before Run 1 CONSTRUCTOR 3
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:**01:35**.5649612 After Run 1 CONSTRUCTOR 3
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:01:35.5720070 After Run 2 CONSTRUCTOR 3
Start - 2022-12-01T15:47:56.6943060-06:00 = 00:01:35.5740155 After Run 3 CONSTRUCTOR 3

2022-12-01T15:47:56.7058479-06:00 Start TEST 2_1
Start - 2022-12-01T15:47:56.7058479-06:00 = 00:00:00.0118613 Before Run 1 TEST 2_1
Start - 2022-12-01T15:47:56.7058479-06:00 = 00:**01:35**.5534193 After Run 1 TEST 2_1
Start - 2022-12-01T15:47:56.7058479-06:00 = 00:01:35.5604651 After Run 2 TEST 2_1
Start - 2022-12-01T15:47:56.7058479-06:00 = 00:01:35.5624736 After Run 3 TEST 2_1

2022-12-01T15:49:32.2683215-06:00 Start TEST 1_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:00:00 Before Run 1 TEST 1_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:**00:00**.0020005 After Run 1 TEST 1_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:00:00.0040005 After Run 2 TEST 1_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:00:00.0060010 After Run 3 TEST 1_2

2022-12-01T15:49:32.2683215-06:00 Start TEST 3_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:00:00 Before Run 1 TEST 3_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:**00:00**.0020005 After Run 1 TEST 3_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:00:00.0040005 After Run 2 TEST 3_2
Start - 2022-12-01T15:49:32.2683215-06:00 = 00:00:00.0060010 After Run 3 TEST 3_2

2022-12-01T15:49:32.2733222-06:00 CONSTRUCTOR 2 (No preload)

2022-12-01T15:49:32.2733222-06:00 Start TEST 2_2
Start - 2022-12-01T15:49:32.2733222-06:00 = 00:00:00 Before Run 1 TEST 2_2
Start - 2022-12-01T15:49:32.2733222-06:00 = 00:**00:00**.0026943 After Run 1 TEST 2_2
Start - 2022-12-01T15:49:32.2733222-06:00 = 00:00:00.0045142 After Run 2 TEST 2_2
Start - 2022-12-01T15:49:32.2733222-06:00 = 00:00:00.0065198 After Run 3 TEST 2_2

2022-12-01T15:49:32.2743225-06:00 Start CONSTRUCTOR 1 (with preload)
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:00:00 Before Run 1 CONSTRUCTOR 1
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:**00:00**.0025146 After Run 1 CONSTRUCTOR 1
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:00:00.0055195 After Run 2 CONSTRUCTOR 1
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:00:00.0095193 After Run 3 CONSTRUCTOR 1

2022-12-01T15:49:32.2743225-06:00 CONSTRUCTOR 3 (with preload)
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:00:00 Before Run 1 CONSTRUCTOR 3
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:**00:00**.0016940 After Run 1 CONSTRUCTOR 3
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:00:00.0055195 After Run 2 CONSTRUCTOR 3
Start - 2022-12-01T15:49:32.2743225-06:00 = 00:00:00.0095193 After Run 3 CONSTRUCTOR 3

2022-12-01T15:49:32.2838418-06:00 Start TEST 1_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:00:00 Before Run 1 TEST 1_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:**00:00**.0045138 After Run 1 TEST 1_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:00:00.0085181 After Run 2 TEST1_ 1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:00:00.0105162 After Run 3 TEST 1_1

2022-12-01T15:49:32.2838418-06:00 Start TEST 3_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:00:00 Before Run 1 TEST 3_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:**00:00**.0045138 After Run 1 TEST 3_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:00:00.0085181 After Run 2 TEST 3_1
Start - 2022-12-01T15:49:32.2838418-06:00 = 00:00:00.0105162 After Run 3 TEST 3_1
```

### Example Code
Three MappingObjectsTestsX classes, identical except No 2 doesn't have the preload in the Constructor.
```
namespace mynamespace
{
    using System;
    using Moq;
    using Xunit;
    using Xunit.Abstractions;

    [Trait("Category", "Mapping.ObjectProperties")]
    public partial class MappingObjectsTests3
    {
        /// <summary>The test output helper.</summary>
        protected readonly ITestOutputHelper TestOutputHelper;

        public MappingObjectsTests3(ITestOutputHelper testOutputHelper)
        {
            TestOutputHelper = testOutputHelper;
            var now = DateTime.Now;
            var start = now;
            TestOutputHelper.WriteLine($"{now:O} CONSTRUCTOR 3 (with preload)");
            var mockingSetup = new MockingSetup { DoAll = true };
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} Before Run 1 CONSTRUCTOR 3");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 1 CONSTRUCTOR 3");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 2 CONSTRUCTOR 3");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 3 CONSTRUCTOR 3");
        }

        [Fact]
        public void VerifyMoqCacheRemembers3_1()
        {
            var now = DateTime.Now;
            var start = now;
            TestOutputHelper.WriteLine($"{now:O} Start TEST 3_1");
            var mockingSetup = new MockingSetup { DoAll = true };
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} Before Run 1 TEST 3_1");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 1 TEST 3_1");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 2 TEST 3_1");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 3 TEST 3_1");
        }

        [Fact]
        public void VerifyMoqCacheRemembers3_2()
        {
            var now = DateTime.Now;
            var start = now;
            TestOutputHelper.WriteLine($"{now:O} Start TEST 3_2");
            var mockingSetup = new MockingSetup { DoAll = true };
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} Before Run 1 TEST 3_2");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 1 TEST 3_2");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 2 TEST 3_2");
            mockingSetup.DoMoqCachingOfTypes();
            now = DateTime.Now;
            TestOutputHelper.WriteLine($"Start - {start:O} = {now - start:c} After Run 3 TEST 3_2");
        }
    }

    public class MockingSetup
    {
        public bool DoAll { get; set; }
        public void DoMoqCachingOfTypes()
        {
             // I have if statement logic around each mock to only run each one if told to
             // DoAll tells it to do all of them so I abbreviated the example to not show the if statements
             _ = new Mock<Account>().Object;
             _ = new Mock<AccountAssociation>().Object;
             _ = new Mock<AccountContact>().Object;
            // ... ~400 classes, each one is an EntityFramework Code First table with properties for
            // columns and relationships
        }
    }
}
```

**Comment by stakx** [comment]
TL;DR: can't be done cleanly on the .NET Framework, and not at all on .NET Core. Some suggestions towards better performance further below.

The type caching is performed by DynamicProxy, the library upon which Moq is built. When running on the .NET Framework (4.x), that library offers a means to save the dynamically generated assembly (which contains the proxy types) to disk, such that it can be loaded back into the type cache on subsequent runs. This feature exists precisely to allow cache pre-warming so the same types don't have to be emitted again and again.

Unfortunately for you, you're migrating to .NET Core soon, which does not support saving dynamic assemblies to disk (it's a System.Reflection.Emit capability that so far has not been ported from the .NET Framework to .NET Core); and even on .NET 4 you cannot use that facility of DynamicProxy because Moq 4 does not expose it. You'd have to use reflection over Moq internals to get at the `ProxyGenerator`. From there you could save/load the type cache.

So yes, you're probably left with your current approach to pre-warm the type cache: loop over all types `T` to be mocked and query for e.g. a throwaway `Mock.Of<T>` to trigger proxy type creation.

You can also somewhat optimize proxy type generation time by keeping the types to be proxied small (e.g. by working with small role interfaces instead of large abstract base classes).

Also, be aware that .NET 5+ are a lot faster than .NET 4, this may somewhat alleviate your performance troubles.

Make sure to update to the latest version of Moq 4, which uses an up-to-date version of DynamicProxy which has some optimizations towards better type reuse. This could also speed up runtime perf significantly.

But finally, with Moq v4, you're at the mercy of System.Reflection.Emit, which is what eventually does all the type generation work. And that is comparatively slow no matter what you do.

Perhaps check out mocking libraries or type proxying libraries that do more work at compile time, such as e.g. the upcoming Moq 5 (see the moq/moq repo), which generates its mock types at compile time using Roslyn source generators.

**Comment by Jothay** [comment]
Thank you for the detailed reply. I will research through your suggestions and try to track down anything I can use towards the end goal
